cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/runtime/distributed/DistributedProgramRunnerTest.java

```diff
@@ -0,0 +1,289 @@
+    Program program = createProgram(baseDir);
+    String runId = RunIds.generate().getId();
+    BasicArguments systemArgs =
+        new BasicArguments(ImmutableMap.of(ProgramOptionConstants.RUN_ID, runId));
+    cConf.set(Constants.AppFabric.PROGRAM_JVM_OPTS, "-Dkey1=val1");
+    BasicArguments userArgs =
+        new BasicArguments(ImmutableMap.of(SystemArguments.JVM_OPTS, "-Dkey2=val2"));
+    ProgramOptions options = new SimpleProgramOptions(program.getId(), systemArgs, userArgs);
+    Mockito.when(twillRunner.prepare(Mockito.any(ProgramTwillApplication.class)))
+        .thenReturn(twillPreparer);
+    Mockito.when(
+            twillPreparer.start(
+                cConf.getLong(Constants.AppFabric.PROGRAM_MAX_START_SECONDS), TimeUnit.SECONDS))
+        .thenReturn(twillController);
+    Mockito.when(twillPreparer.withApplicationClassPaths(Mockito.any(Iterable.class)))
+        .thenReturn(twillPreparer);
+    Mockito.when(twillPreparer.withBundlerClassAcceptor(Mockito.any())).thenReturn(twillPreparer);
+    Mockito.when(
+            twillPreparer.withApplicationArguments(
+                DistributedProgramRunner.PROGRAM_OPTIONS_FILE_NAME))
+        .thenReturn(twillPreparer);
+    Mockito.doAnswer(
+            i -> {
+              Runnable runnable = i.getArgumentAt(0, Runnable.class);
+              Executor executor = i.getArgumentAt(1, Executor.class);
+              executor.execute(runnable);
+              return null;
+            })
+        .when(twillController)
+        .onRunning(Mockito.any(Runnable.class), Mockito.eq(Threads.SAME_THREAD_EXECUTOR));
+
+    ProgramController programController = runner.run(program, options);
+
+    ProgramRunId expectedProgramRunId =
+        new ProgramRunId(Id.Namespace.DEFAULT.getId(), "app", ProgramType.SPARK, "spark", runId);
+    Assert.assertEquals(expectedProgramRunId, programController.getProgramRunId());
+    Mockito.verify(twillPreparer)
+        .withConfiguration(
+            ImmutableMap.of(
+                Configs.Keys.YARN_MAX_APP_ATTEMPTS,
+                "1",
+                ProgramOptionConstants.RUNTIME_NAMESPACE,
+                program.getNamespaceId()));
+    Mockito.verify(twillPreparer).addJVMOptions("-Dkey1=val1 -Dkey2=val2");
+    Mockito.verify(twillPreparer)
+        .withEnv(
+            Collections.singletonMap("CDAP_LOG_DIR", ApplicationConstants.LOG_DIR_EXPANSION_VAR));
+    Mockito.verify(twillPreparer).withEnv(extraEnv);
+    Mockito.verify(twillPreparer)
+        .withApplicationArguments(DistributedProgramRunner.PROGRAM_OPTIONS_FILE_NAME);
+    Mockito.verify(twillPreparer).setClassLoader(MainClassLoader.class.getName());
+    List<String> yarnAppClassPath =
+        Arrays.asList(
+            hConf.getTrimmedStrings(
+                YarnConfiguration.YARN_APPLICATION_CLASSPATH,
+                YarnConfiguration.DEFAULT_YARN_APPLICATION_CLASSPATH));
+    Mockito.verify(twillPreparer).withApplicationClassPaths(yarnAppClassPath);
```

Comment:
this type of test is brittle because it is essentially testing the internal implementation of the DistributedProgramRunner and not the actual contract defined by the ProgramRunner methods. Changing the implementation will often require changing this test. Using Mockito for our own internal classes tends to have this sort of problem, it is usually more suitable to mock classes that we don't have control over. Normally it is better to try and structure the code in a more testable way (part of the difficulty is that DistributedProgramRunner.run() is one long method) rather than doing something like this.

Is there something more specific that we are trying to test, like the jvm options?

Reply:
Yeah, the intention was to test the jvm options.

Reply:
Refactored DistributedProgramRunner, added tests for smaller methods. Also removed mocks